File: neutrino-binding/src/main/java/de/hhu/bsinfo/observatory/neutrino/ConnectionInformation.java
package de.hhu.bsinfo.observatory.neutrino;

import java.nio.ByteBuffer;
import java.util.StringJoiner;

public class ConnectionInformation {

    private final byte portNumber;
    private final short localId;
    private final int queuePairNumber;

    ConnectionInformation(byte portNumber, short localId, int queuePairNumber) {
        this.portNumber = portNumber;
        this.localId = localId;
        this.queuePairNumber = queuePairNumber;
    }

    ConnectionInformation(ByteBuffer buffer) {
        portNumber = buffer.get();
        localId = buffer.getShort();
        queuePairNumber = buffer.getInt();
    }

    static int getSizeInBytes() {
        return Byte.BYTES + Short.BYTES + Integer.BYTES;
    }

    static ConnectionInformation fromBytes(byte[] bytes) {
        ByteBuffer buffer = ByteBuffer.wrap(bytes);

        return new ConnectionInformation(buffer.get(), buffer.getShort(), buffer.getInt());
    }

    byte getPortNumber() {
        return portNumber;
    }

    short getLocalId() {
        return localId;
    }

    int getQueuePairNumber() {
        return queuePairNumber;
    }

    byte[] toBytes() {
        ByteBuffer buffer = ByteBuffer.allocate(getSizeInBytes());

        buffer.put(portNumber);
        buffer.putShort(localId);
        buffer.putInt(queuePairNumber);

        return buffer.array();
    }

    @Override
    public String toString() {
        return "ConnectionInformation {" +
                "\n\tportNumber=" + portNumber +
                ",\n\tlocalId=" + localId +
                ",\n\tqueuePairNumber=" + queuePairNumber +
                "\n}";
    }
}

File: cpp/src/bindings/verbs-binding/ConnectionInformation.cpp
#include <sstream>
#include "ConnectionInformation.h"

namespace Verbs {

ConnectionInformation::ConnectionInformation(uint8_t portNumber, uint16_t localId, uint32_t queuePairNumber) :
        portNumber(portNumber),
        localId(localId),
        queuePairNumber(queuePairNumber) {}

ConnectionInformation ConnectionInformation::fromBytes(const uint8_t *bytes) {
    uint8_t portNumber = *bytes;
    bytes += sizeof(uint8_t);

    uint16_t localId = *reinterpret_cast<const uint16_t*>(bytes);
    bytes += sizeof(uint16_t);

    uint32_t queuePairNumber = *reinterpret_cast<const uint32_t*>(bytes);

    return ConnectionInformation(portNumber, localId, queuePairNumber);
}

std::shared_ptr<uint8_t[]> ConnectionInformation::toBytes() {
    auto bytes = std::shared_ptr<uint8_t[]>(new uint8_t[sizeof(ConnectionInformation)]);
    uint8_t index = 0;

    *bytes.get() = portNumber;
    index += sizeof(uint8_t);

    *reinterpret_cast<uint16_t*>(&bytes.get()[index]) = localId;
    index += sizeof(uint16_t);

    *reinterpret_cast<uint32_t*>(&bytes.get()[index]) = queuePairNumber;

    return bytes;
}

uint8_t ConnectionInformation::getPortNumber() {
    return portNumber;
}

uint16_t ConnectionInformation::getLocalId() {
    return localId;
}

uint32_t ConnectionInformation::getQueuePairNumber() {
    return queuePairNumber;
}

ConnectionInformation::operator std::string() {
    std::ostringstream stream;

    stream << "ConnectionInformation {"
           << "\n\t" << "portNumber=" << static_cast<uint16_t>(portNumber)
           << ",\n\t" << "localId=" << localId
           << ",\n\t" << "queuePairNumber=" << queuePairNumber
           << "\n}";

    return stream.str();
}

}